src/libre/re_ast.h

```diff
@@ -90,34 +131,52 @@ struct ast_expr {
 		struct {
 			enum re_flags pos;
 			enum re_flags neg;
-			/* Previous flags, saved here and restored
-			 * when done compiling the flags node's subtree.
-			 * Since `struct ast_expr` contains a union,
-			 * this space is already allocated. */
-			enum re_flags saved;
 		} flags;
+		struct {
+			enum re_ast_anchor_type t;
+		} anchor;
```

Comment:
I'm surprised the anchors are independent nodes, rather than a bitmap attached to every node. Did you reject that idea already, or is it worth exploring?

Reply:
I spent at least a week and a half on this specific issue.

My first several attempts converted them to flags on adjacent nodes, but that leads to some incredibly tricky cases when there are floating anchors, like `/(^)abc/`, or redundant but satisfiable anchors, like `/^^^abc/`.

Keeping them as independent nodes meant that they could behave like EMPTY nodes (x ---epsilon---> y), except the start or end is the global start or end, and then the epsilon closure algorithm during determinization already handles repeated satisfiable anchors correctly. Eventually, I figured that I was duplicating analysis it was doing later anyway.

It still needs to scan for unsatisfiable anchors (such as `/^a^bc/`, where it has non-nullable input before a non-nullable start anchor, and the corresponding end anchor case `/^ab$c$/`), because otherwise it ends up with false positives. Without explicit pruning, those would falsely match "bc" and "ab" respectively.